src/drivers/intel/ssp/ssp.c

```diff
@@ -274,6 +274,14 @@ static int ssp_set_config(struct dai *dai, struct ipc_config_dai *common_config,
 		goto out;
 	}
 
+	if ((config->ssp.clks_control & SOF_DAI_INTEL_SSP_CLKCTRL_MCLK_ES ||
+	     config->ssp.clks_control & SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES) &&
+	    !(config->flags & SOF_DAI_CONFIG_FLAGS_HW_PARAMS ||
+	      config->flags & SOF_DAI_CONFIG_FLAGS_HW_FREE)) {
+		dai_info(dai, "ssp_set_config(): ES bits set but hw_params or hw_free bits not set. Ignore config");
+		goto out;
```

Comment:
Is the test necessary? IMHO, we should not send DAI_CONFIG instead of ignoring it within ssp_set_config() if a DAI_CONFIG is not needed. We will not ignore config when clks_control is not set. That say the calling from widget setup will be ignored only when clks_control is set and I don't think we should have different behavior except the clock configuration between `early start flag` is set or not.

Reply:
I think the warning is always useful when figuring out any issues. so this should maybe be a dai_warn();

Reply:
@bardliao the problem is that the code for the DAI_CONFIG is generic on the kernel side, there's nothing that the SOF driver can look for to know if the DAI will be configured again later in the .hw_params state.

Currently only HDaudio, SoundWire and SSP (since the multi-config work) are configured during the .hw_params or prepare stages. DMIC is only configured on startup (or on pipeline enablement). I don't want to make any assumptions on what NXP are doing.

To avoid sending the DAI_CONFIG, I would need to add yet another token and/more flags. I took the simpler approach of just ignoring the spurious IPC. It's not super elegant but it avoids the need for more topology changes.

Reply:
> I think the warning is always useful when figuring out any issues. so this should maybe be a dai_warn();

It's not really a warning people need to worry about, this message will be shown in all SSP uses when the initial DAI_CONFIG is sent. I just wanted to make sure people can correlate the DAI_CONFIGs sent by the kernel to what happens in the firmware.

Reply:
test removed.